### PR TITLE
added locals to beforeInstall/beforeUninstall

### PR DESCRIPTION
**Pull request by mattmarcum**
updated the docs to document this, and added unit tests with updated basicBlueprint fixtures

### CODE REVIEW

**Comment by mattmarcum** [comment]
This addresses https://github.com/ember-cli/ember-cli/issues/3146

at the time I created this master branch was not passing tests locally for me, but ci says it is...

**Comment by jgwhite** [comment]
This is looking good. I’m a little hesitant about mutating `locals` in-place (rather than returning a new set) mainly because it’s a pattern we’re not using elsewhere. That said, it looks like the most practical solution in this case.

@trabus can we get your eyes on this too?

File: tests/unit/models/blueprint-test.js [hunk 1]
```diff
@@ -225,7 +226,7 @@ describe('Blueprint', function() {
         .then(function() {
           var actualFiles = walkSync(tmpdir).sort();
           var output = ui.output.trim().split(EOL);
-
+
```
Comment (jgwhite) [inline tests/unit/models/blueprint-test.js]: Bit of trailing whitespace here :P

**Comment by mattmarcum** [comment]
I had started to create the pattern where locals were returned like this:

```javascript
return Promise.resolve()
    .then(function(){ 
      return this.locals(options); 
    }.bind(this))
    .then(function(customLocals){ 
      return this._locals(options, customLocals);
    }.bind(this))
    .then(function(locals){
      var promise = this.beforeInstall(options);
      if(promise && promise.then){
        return promise.then(function(){
          return locals;
        });
      }else{
        return locals;
      }
    }.bind(this))
    .then(function(locals){
      return this.processFiles(intoDir, locals);
    }.bind(this))
    .map(commit)
    .then(this.afterInstall.bind(this, options));
```
But this doesn't help if you want to do async work in fileMapTokens too.  The call to that function is several functions deep off of _locals.  It would take a major refactor to be able to handle a promise returned from fileMapTokens.  I think with the unit test here, we would know if `locals` ever gets removed from the scope of the before(Un)Install functions.

**Comment by mattmarcum** [comment]
The smoke test failed because of a timeoute...they were passing earlier and then I removed some whitespace.  I bet if they are run again they will pass.

**Comment by mattmarcum** [comment]
slow-tests are still timing out, but now on a different test

**Comment by stefanpenner** [comment]
rebuilding...

**Comment by trabus** [comment]
I think this is okay for now, but I like the idea of making the locals and fileMapTokens hooks promise aware as a longterm fix.

**Comment by mattmarcum** [comment]
Does anything need to be changed on the pr before it can be merged to master?

**Comment by jgwhite** [comment]
I’m ready to push the button on this. @mattmarcum could you work up a PR on `gh-pages` updating the docs for `beforeInstall`?

**Comment by jgwhite** [comment]
Note to self: make sure pull requests are squashed before merge :)

**Pull request merged**